doc/SAI-Proposal-Debug-Counters.md

```diff
@@ -0,0 +1,264 @@
+{
+    /* L2 reasons */
+	
+    /** Source MAC is multicast */
+    SAI_PORT_IN_DROP_REASON_SMAC_MULTICAST,
+
+    /** Source MAC equals Destination MAC */
+    SAI_PORT_IN_DROP_REASON_SMAC_EQUALS_DMAC,
+
+    /** Destination MAC is Reserved (DMAC=01-80-C2-00-00-0x) */
+    SAI_PORT_IN_DROP_REASON_DMAC_RESERVED,
+
+    /** 
+     * @brief VLAN tag not allowed
+     *
+     * Frame tagged when port is dropping tagged, 
+     * or untagged when dropping untagged
+     */
+    SAI_PORT_IN_DROP_REASON_VLAN_TAG_NOT_ALLOWED,
+
+    /** Ingress VLAN filter */
+    SAI_PORT_IN_DROP_REASON_INGRESS_VLAN_FILTER,
+
+    /** Ingress STP filter */
+    SAI_PORT_IN_DROP_REASON_INGRESS_STP_FILTER,
+	
+    /** Unicast FDB table action discard */
+    SAI_PORT_IN_DROP_REASON_FDB_UC_DISCARD,
+
+    /** Multicast FDB table empty tx list */
+    SAI_PORT_IN_DROP_REASON_FDB_MC_DISCARD,
+
+    /** Port loopback filter */
+    SAI_PORT_IN_DROP_REASON_LOOPBACK_FILTER,
+
+    /* L3 reasons */
+	
+    /** IPv4 Unicast Destination IP is link local (Destination IP=169.254.0.0/16) */
+    SAI_PORT_IN_DROP_REASON_DIP_LINK_LOCAL,
+
+    /** IPv4 Source IP is link local (Source IP=169.254.0.0/16) */
+    SAI_PORT_IN_DROP_REASON_SIP_LINK_LOCAL,
+
+    /** Packet size is larger than the MTU */
+    SAI_PORT_IN_DROP_REASON_EXCEEDS_MTU,
+
+    /* ACL reasons */
+
+    /** Packet is dropped due to configured ACL rules */
+    SAI_PORT_IN_DROP_REASON_ACL_DISCARD	
+
+} sai_port_in_drop_reason_t;
+```
+
+## Debug counter port out drop reason family
+```
+typedef enum _sai_port_out_drop_reason_t
+{
+    /** Egress VLAN filter */
+    SAI_PORT_OUT_DROP_REASON_EGRESS_VLAN_FILTER,
+	
+    /** Packet is destined for neighboring device but neighbor device link is down */
+    SAI_PORT_OUT_DROP_REASON_L3_EGRESS_LINK_DOWN,	
+
+} sai_port_out_drop_reason_t;
+```
+
+### Configuring a debug counter
+```
+/**
+ * @brief Attribute Id in sai_set_counter_attribute() and
+ * sai_get_counter_attribute() calls
+ */
+typedef enum _sai_debug_counter_attr_t
+{
+    /**
+     * @brief Start of attributes
+     */
+    SAI_DEBUG_COUNTER_ATTR_START,
+
+    /* READ-ONLY */
+
+    /**
+     * @brief Object stat index
+     * Index is added to base start
+     *
+     * @type sai_uint32_t
+     * @flags READ_ONLY
+     */
+    SAI_DEBUG_COUNTER_ATTR_INDEX = SAI_DEBUG_COUNTER_ATTR_START,
+	
+    /* READ-WRITE */
+
+    /**
+     * @brief Debug counter type
+     *
+     * @type sai_debug_counter_type_t
+     * @flags MANDATORY_ON_CREATE | CREATE_ONLY
+     * @isresourcetype true
+     */
+    SAI_DEBUG_COUNTER_ATTR_TYPE,
+
+    /**
+     * @brief Bind method to base object
+     *
+     * @type sai_debug_counter_bind_method_t
+     * @flags CREATE_ONLY
+     * @default SAI_DEBUG_COUNTER_BIND_METHOD_AUTOMATIC
+     */
+    SAI_DEBUG_COUNTER_ATTR_BIND_METHOD,	
+
+    /**
+     * @brief List of port in drop reasons that will be counted
+     *
+     * @type sai_s32_list_t sai_port_in_drop_reason_t
```

Comment:
Just to be explicit. When a packet is deemed to be 'eligible' for a drop, it can be for more than one drop reason. In that sense, this debug-counter is counting the number of times a given drop-reason has come into play, as opposed to number of packets that have been dropped. So this counter, for a given port, would be greater-than-or-equal to the number of packets dropped on that port (assuming a drop at INGRESS as an example). 

It would be good to document this behaviour in the MD file.

Reply:
This is interesting. If the drop reasons are not meant to be mutually exclusive, then we can add few more drop reasons at a coarser granularity. For eg, along with the proposed fine grained drop reasons, we could have coarser drop-reasons like SAI_PORT_IN_DROP_REASON_L2_ANY, SAI_PORT_IN_DROP_REASON_L3_ANY etc. 
Use cases:
1) It allows NPU/ASIC vendors to accommodate ASIC specific drop counters which are not listed in the SAI spec.
2) As a user when I want to debug the drop reason for a given flow, I would first use the coarse grained counters like L2 or L3. Based on which of the counter got incremented, I can drill down to the other fine-grained counters.

Reply:
Added a section about counting a packet dropped because of multiple reasons in the MD
Packet will be counted once per debug counter instance

Added drop reason any L2, L3 as suggested